### DISCUSSION TITLE
Programatically add new donor sans donation

### DISCUSSION THREAD

**Discussion by craig02445** [impress-org/givewp#6814] [Developer Q&A] [answered]
I am looking to programmatically add a couple of donors who have not yet donated. I believe I need this function [create](https://github.com/impress-org/givewp/blob/master/src/Donors/Models/Donor.php#L123), is that the only one I need? Can I get a simple example of calling it from my own form? I can create the array, just not sure of the syntax of calling the function from my own snippet. Thanks!
Also, where can I find the source code for the IMPORT tool / function?

**Comment by jonwaldstein** [comment, marked as answer]
Yup you can use the Donor model to create a new one like this:
```
$donor = Give\Donors\Models\Donor::create([
    'name' => trim("$firstName $lastName"),
    'firstName' => $firstName,
    'lastName' => $lastName,
    'email' => $donorEmail,
    'userId' => $userId ?: null
]);

// once created, it is inserted into the database and you can access the ID and other properties.
$donor->id
```

**Reply by craig02445** [reply]
Perfect, THANK YOU.

**Reply by craig02445** [reply]
What happens if I do this call and the donor already exists?
And I guess the other question is, what is the easiest way to see if a donor already exists?
THANKS.

**Comment by kjohnson** [comment]
> Also, where can I find the source code for the IMPORT tool / function?

I wasn't sure myself, so I did a quick dive to document it here, but I would recommend using the new Donor Model factory that @jonwaldstein mentions.

It starts in the [Give_Import_Donations](includes/admin/tools/import/class-give-import-donations.php), which triggers [give_donation_import_callback](includes/admin/admin-actions.php), which calls [give_save_import_donation_to_db](includes/admin/import-functions.php), which finally calls [creates a new donor](
https://github.com/impress-org/givewp/blob/b7ec45e0aea0bd40fb5e434aedb4330c7c31ef08/includes/admin/import-functions.php#L320).

**Reply by craig02445** [reply]
Great, thanks for the help.
